hazelcast/src/test/java/com/hazelcast/sql/impl/worker/QueryFragmentExecutableTest.java

```diff
@@ -0,0 +1,478 @@
+
+            AtomicInteger doneOperationCount = new AtomicInteger();
+            CountDownLatch doneLatch = new CountDownLatch(threadCount);
+
+            Runnable run = () -> {
+                try {
+                    while (true) {
+                        int counter = doneOperationCount.getAndIncrement();
+
+                        if (counter >= operationCount) {
+                            break;
+                        }
+
+                        QueryAbstractExchangeOperation operation;
+
+                        if (ThreadLocalRandom.current().nextBoolean()) {
+                            operation = new QueryBatchExchangeOperation(queryId, edgeId, EmptyRowBatch.INSTANCE, false, counter);
+                        } else {
+                            operation = new QueryFlowControlExchangeOperation(queryId, edgeId, counter);
+                        }
+
+                        operation.setCallerId(callerId);
+
+                        fragmentExecutable.addOperation(operation);
+                        fragmentExecutable.schedule();
+                    }
+                } finally {
+                    doneLatch.countDown();
+                }
+            };
+
+            for (int j = 0; j < threadCount; j++) {
+                new Thread(run).start();
+            }
+
+            doneLatch.await();
+
+            // Make sure that all batches were drained.
+            assertTrueEventually(() -> assertTrue(inboundQueue.isEmpty()));
+            assertTrueEventually(() -> assertTrue(outboundQueue.isEmpty()));
+
+            // Make sure that the executor observed all batches.
+            assertTrueEventually(() -> assertEquals(operationCount, inboundSet.size() + outboundSet.size()));
+        }
+    }
+
+    /**
+     * Test that ensures propagation of cancel.
+     */
+    @Test
+    public void testShutdown() throws Exception {
+        pool = createPool();
+
+        TestStateCallback stateCallback = new TestStateCallback();
+        TestExec exec = new TestExec();
+
+        QueryFragmentExecutable fragmentExecutable = new QueryFragmentExecutable(
+            stateCallback,
+            Collections.emptyList(),
+            exec,
+            Collections.emptyMap(),
+            Collections.emptyMap(),
+            pool
+        );
+
+        CountDownLatch startLatch = new CountDownLatch(1);
+        CountDownLatch blockLatch = new CountDownLatch(1);
+        AtomicBoolean interruptCaught = new AtomicBoolean();
+
+        exec.setPayload(() -> {
+            startLatch.countDown();
+
+            try {
+                blockLatch.await();
+            } catch (InterruptedException e) {
+                interruptCaught.set(true);
+
+                Thread.currentThread().interrupt();
+
+                throw e;
+            }
+
+            return IterationResult.FETCHED;
+        });
+
+        // Await exec is reached.
+        assertTrue(fragmentExecutable.schedule());
+        startLatch.await();
```

Comment:
The same

Reply:
Likewise, this latch doesn't test any assumptions. Instead, it just positions the threads at proper source code lines.